docs/rfcs/025-sk-peer-recovery.md

```diff
@@ -0,0 +1,71 @@
+# Safekeeper peer recovery
+
+## A problem and proposed solution
+
+Currently, on start walproposer preserves WAL for all safekeepers: on start it
+determines the horizon beyond which all safekeepers received WAL, downloads
+missing part to pg_wal and holds WAL since this horizon ever since. This is
+problematic, because
+
+1. If one safekeeper is down and/or lagging, pg_wal eventually explodes -- we intentionally don't have much space on computes.
+2. If one safekeeper is down and/or lagging it makes compute start longer.
+
+Proposed solution is to teach safekeepers to fetch WAL directly from peers,
+respecting consensus rules. Namely,
+- On start, walproposer won't download WAL at all -- it will have it only since
```

Comment:
Not much changes here, sync-safekeepers would wait for commit_lsn reaching `propEpochStartLsn` in the same way. There is a question how walproposer should handle the case when it doesn't have WAL for a particular safekeeper. The simplest way seems to be to check for this during each read and send ERROR to the safekeeper if WAL is missing, logging and closing the connection. Then during recovery walproposer -> sk communications would look like connect, vote, send ProposerElected, log error 'no WAL' with lsns etc, rinsing and repeating each second.

Alternatively we could add special state kinda 'waiting' in the protocol, but it is harder and isn't much useful.

So an addition here is 'lowest available LSN' variable tracked in walproposer, it seems we don't have it currently.

Reply:
> Then during recovery walproposer -> sk communications would look like connect, vote, send ProposerElected, log error 'no WAL' with lsns etc, rinsing and repeating each second.

Then it means that sync-safekeepers will work like this, retrying connection every second until 2/3 safekeepers will reach `propEpochStartLsn`. If peer recovery will align flushLsn on safekeepers, it will work fine, but otherwise (if peer recovery condition is not met, for example), sync-safekeepers will work for at least a second.

Reply:
Yes. But practically it should almost never happen, since aligning should happen in the background even in compute absense.